File: app/src/main/java/info/nightscout/androidaps/plugins/ProfileNS/NSProfileFragment.java
package info.nightscout.androidaps.plugins.ProfileNS;

import android.app.Activity;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.Spinner;
import android.widget.TextView;

import com.crashlytics.android.Crashlytics;
import com.squareup.otto.Subscribe;

import java.util.ArrayList;

import info.nightscout.androidaps.MainApp;
import info.nightscout.androidaps.R;
import info.nightscout.androidaps.data.Profile;
import info.nightscout.androidaps.data.ProfileStore;
import info.nightscout.androidaps.plugins.Common.SubscriberFragment;
import info.nightscout.androidaps.plugins.ProfileNS.events.EventNSProfileUpdateGUI;
import info.nightscout.utils.DecimalFormatter;


public class NSProfileFragment extends SubscriberFragment implements AdapterView.OnItemSelectedListener {
    private Spinner profileSpinner;
    private TextView noProfile;
    private TextView units;
    private TextView dia;
    private TextView activeProfile;
    private TextView ic;
    private TextView isf;
    private TextView basal;
    private TextView target;

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        try {
            View layout = inflater.inflate(R.layout.nsprofile_fragment, container, false);

            profileSpinner = (Spinner) layout.findViewById(R.id.nsprofile_spinner);
            noProfile = (TextView) layout.findViewById(R.id.profileview_noprofile);
            units = (TextView) layout.findViewById(R.id.profileview_units);
            dia = (TextView) layout.findViewById(R.id.profileview_dia);
            activeProfile = (TextView) layout.findViewById(R.id.profileview_activeprofile);
            ic = (TextView) layout.findViewById(R.id.profileview_ic);
            isf = (TextView) layout.findViewById(R.id.profileview_isf);
            basal = (TextView) layout.findViewById(R.id.profileview_basal);
            target = (TextView) layout.findViewById(R.id.profileview_target);

            profileSpinner.setOnItemSelectedListener(this);

            updateGUI();
            return layout;
        } catch (Exception e) {
            Crashlytics.logException(e);
        }

        return null;
    }

    @Subscribe
    public void onStatusEvent(final EventNSProfileUpdateGUI ev) {
        Activity activity = getActivity();
        if (activity != null)
            activity.runOnUiThread(new Runnable() {
                @Override
                public void run() {
                    updateGUI();
                }
            });
    }

    @Override
    protected void updateGUI() {
        if (MainApp.getConfigBuilder().getProfile() == null) {
            noProfile.setVisibility(View.VISIBLE);
            return;
        } else {
            noProfile.setVisibility(View.GONE);
        }

        ProfileStore profileStore = NSProfilePlugin.getPlugin().getProfile();
        if (profileStore != null) {
            ArrayList<CharSequence> profileList = profileStore.getProfileList();
            ArrayAdapter<CharSequence> adapter = new ArrayAdapter<>(getContext(),
                    R.layout.spinner_centered, profileList);
            profileSpinner.setAdapter(adapter);
            // set selected to actual profile
            for (int p = 0; p < profileList.size(); p++) {
                if (profileList.get(p).equals(MainApp.getConfigBuilder().getProfileName()))
                    profileSpinner.setSelection(p);
            }
        }
    }

    @Override
    public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
        String name = parent.getItemAtPosition(position).toString();

        ProfileStore store = NSProfilePlugin.getPlugin().getProfile();
        if (store != null) {
            Profile profile = store.getSpecificProfile(name);
            if (profile != null) {
                units.setText(profile.getUnits());
                dia.setText(DecimalFormatter.to2Decimal(profile.getDia()) + " h");
                activeProfile.setText(name);
                ic.setText(profile.getIcList());
                isf.setText(profile.getIsfList());
                basal.setText(profile.getBasalList());
                target.setText(profile.getTargetList());
            }
        }
    }

    @Override
    public void onNothingSelected(AdapterView<?> parent) {
        noProfile.setVisibility(View.VISIBLE);
        units.setText("");
        dia.setText("");
        activeProfile.setText("");
        ic.setText("");
        isf.setText("");
        basal.setText("");
        target.setText("");
    }
}

File: app/src/main/java/info/nightscout/androidaps/plugins/profile/ns/NSProfileFragment.kt
package info.nightscout.androidaps.plugins.profile.ns

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.AdapterView
import android.widget.ArrayAdapter
import dagger.android.support.DaggerFragment
import info.nightscout.androidaps.R
import info.nightscout.androidaps.databinding.NsprofileFragmentBinding
import info.nightscout.androidaps.interfaces.ProfileFunction
import info.nightscout.androidaps.plugins.bus.RxBusWrapper
import info.nightscout.androidaps.plugins.profile.ns.events.EventNSProfileUpdateGUI
import info.nightscout.androidaps.plugins.treatments.TreatmentsPlugin
import info.nightscout.androidaps.utils.DateUtil
import info.nightscout.androidaps.utils.DecimalFormatter
import info.nightscout.androidaps.utils.FabricPrivacy
import info.nightscout.androidaps.utils.alertDialogs.OKDialog
import info.nightscout.androidaps.utils.extensions.plusAssign
import info.nightscout.androidaps.utils.resources.ResourceHelper
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import javax.inject.Inject

class NSProfileFragment : DaggerFragment() {

    @Inject lateinit var treatmentsPlugin: TreatmentsPlugin
    @Inject lateinit var rxBus: RxBusWrapper
    @Inject lateinit var resourceHelper: ResourceHelper
    @Inject lateinit var fabricPrivacy: FabricPrivacy
    @Inject lateinit var profileFunction: ProfileFunction
    @Inject lateinit var nsProfilePlugin: NSProfilePlugin

    private var disposable: CompositeDisposable = CompositeDisposable()

    private var _binding: NsprofileFragmentBinding? = null

    // This property is only valid between onCreateView and
    // onDestroyView.
    private val binding get() = _binding!!

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View {
        _binding = NsprofileFragmentBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        binding.profileviewer.closeLayout.close.visibility = View.GONE // not needed for fragment

        binding.profileswitch.setOnClickListener {
            val name = binding.spinner.selectedItem?.toString() ?: ""
            nsProfilePlugin.profile?.let { store ->
                store.getSpecificProfile(name)?.let {
                    activity?.let { activity ->
                        OKDialog.showConfirmation(activity, resourceHelper.gs(R.string.nsprofile),
                            resourceHelper.gs(R.string.activate_profile) + ": " + name + " ?", Runnable {
                            treatmentsPlugin.doProfileSwitch(store, name, 0, 100, 0, DateUtil.now())
                        })
                    }
                }
            }
        }

        binding.spinner.onItemSelectedListener = object : AdapterView.OnItemSelectedListener {
            override fun onNothingSelected(parent: AdapterView<*>?) {
                if (_binding == null) return
                binding.profileviewer.invalidprofile.visibility = View.VISIBLE
                binding.profileviewer.noprofile.visibility = View.VISIBLE
                binding.profileviewer.units.text = ""
                binding.profileviewer.dia.text = ""
                binding.profileviewer.activeprofile.text = ""
                binding.profileviewer.ic.text = ""
                binding.profileviewer.isf.text = ""
                binding.profileviewer.basal.text = ""
                binding.profileviewer.basaltotal.text = ""
                binding.profileviewer.target.text = ""
                binding.profileswitch.visibility = View.GONE
            }

            override fun onItemSelected(parent: AdapterView<*>?, view: View?, position: Int, id: Long) {
                if (_binding == null) return
                val name = binding.spinner.getItemAtPosition(position).toString()

                binding.profileswitch.visibility = View.GONE

                nsProfilePlugin.profile?.let { store ->
                    store.getSpecificProfile(name)?.let { profile ->
                        if (_binding == null) return
                        binding.profileviewer.units.text = profile.units
                        binding.profileviewer.dia.text = resourceHelper.gs(R.string.format_hours, profile.dia)
                        binding.profileviewer.activeprofile.text = name
                        binding.profileviewer.ic.text = profile.icList
                        binding.profileviewer.isf.text = profile.isfList
                        binding.profileviewer.basal.text = profile.basalList
                        binding.profileviewer.basaltotal.text = String.format(resourceHelper.gs(R.string.profile_total), DecimalFormatter.to2Decimal(profile.baseBasalSum()))
                        binding.profileviewer.target.text = profile.targetList
                        binding.profileviewer.basalGraph.show(profile)
                        if (profile.isValid("NSProfileFragment")) {
                            binding.profileviewer.invalidprofile.visibility = View.GONE
                            binding.profileswitch.visibility = View.VISIBLE
                        } else {
                            binding.profileviewer.invalidprofile.visibility = View.VISIBLE
                            binding.profileswitch.visibility = View.GONE
                        }
                    }
                }
            }
        }
    }

    @Synchronized
    override fun onResume() {
        super.onResume()
        disposable += rxBus
            .toObservable(EventNSProfileUpdateGUI::class.java)
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe({ updateGUI() }, { fabricPrivacy.logException(it) })
        updateGUI()
    }

    @Synchronized
    override fun onPause() {
        super.onPause()
        disposable.clear()
    }

    @Synchronized
    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }

    @Synchronized
    fun updateGUI() {
        if (_binding == null) return
        binding.profileviewer.noprofile.visibility = View.VISIBLE

        nsProfilePlugin.profile?.let { profileStore ->
            context?.let { context ->
                val profileList = profileStore.getProfileList()
                val adapter = ArrayAdapter(context, R.layout.spinner_centered, profileList)
                binding.spinner.adapter = adapter
                // set selected to actual profile
                for (p in profileList.indices) {
                    if (profileList[p] == profileFunction.getProfileName())
                        binding.spinner.setSelection(p)
                }
                binding.profileviewer.noprofile.visibility = View.GONE
            }
        }
    }
}
